### ISSUE TITLE
Python libraries not recompiled by make after changes to OPM source file

### ISSUE THREAD

**Issue by hakonhagland** [OPM/opm-common#2823]
I am not able to rebuild the Python libraries in `opm-common` (using `make` in the `build` folder) after having changed an OPM source file. Here is a reproducible minimal example:
```
git clone git@github.com:OPM/opm-common.git
cd opm-common
mkdir build
cd build
cmake -DOPM_ENABLE_PYTHON=ON ..
make -j12
```
Then create a test script in the `python` build folder:
```
$ cd python
# create mytest.py file like this:
$ cat mytest.py
from opm.io.parser import Builtin
from opm.io.parser import Parser
from opm.io.parser import ParseContext
from opm.io.deck import DeckKeyword

parser = Parser()
data = """
START
10 MAI 2007 /
RUNSPEC
FIELD
DIMENS
2 2 1 /
GRID
DX
4*0.25 /
DY
4*0.25 /
DZ
4*0.25 /
"""
deck = parser.parse_string(data)
active_unit_system = deck.active_unit_system()
default_unit_system = deck.default_unit_system()
kw = DeckKeyword(
    parser["AQUCWFAC"], [["*", 0.25]], active_unit_system, default_unit_system)
```
Run the test script:
```
$ python3 mytest.py
# No output (as expected)
```
Change an OPM source file that is used by the test script, for example `DeckKeyword.cpp`, add the line:
```
std::cout << "processing " << parserKeyword.getName() << '\n';
```
at line 94:
```
sed -i '94i std::cout << "processing " << parserKeyword.getName() << "\\n";' ../../src/opm/parser/eclipse/Deck/DeckKeyword.cpp
```
Rerun `make`:
```
cd ..
make -j12
```
Rerun the test script:
```
cd python
$ python3 mytest.py
# Still no output (not expected)
```
Remake all targets:
```
cd ..
make -B -j12
```
If I now rerun the test script, I get the expected output:
```
cd python
$ python3 mytest.py
processing AQUCWFAC
```

Is this a bug in the `Makefile`? How can I recompile the Python libraries without having to rebuild all targets?

**Comment by blattms** [comment]
I don't think these dependencies can be seen by Cmake automatically. One has to tell it manually via `add_dependencies(target dep1 dep2 ...)`, e,g. that the python stuff/lib depends on the opmcommon libe

**Comment by joakim-hove** [comment]
> I don't think these dependencies can be seen by Cmake automatically. One has to tell it manually via add_dependencies(target dep1 dep2 ...), e,g. that the python stuff/lib depends on the opmcommon lib

Yes - that is right; this dependency is not natively seen by `cmake`, we have tried to configure that dependency manually with `add_depenencies(....) but that obviously does not work to perfection - and I quite frankly do not understand it myself when I look at the code.

What I think would solve your immediate problem was to build the main opmcommon library as a shared library, then the final linking of the python library is deferred to runtime:

1. Make sure to remove static libraries from `${build}/lib`
2. Configure cmake with `cmake . -DBUILD_SHARED_LIBS=ON`
3. Recompile

Then you should get the functionality you desire - although it is through a workaround.

**Comment by hakonhagland** [comment]
> What I think would solve your immediate problem was to build the main opmcommon library as a shared library

@joakim-hove Good idea. I tried this now, but it seem like the linker does not find the python shared library:

```
$ mkdir build2
$ cmake -DOPM_ENABLE_PYTHON=ON -DBUILD_SHARED_LIBS=ON ..
$ make -j12
[...]
[ 55%] Building CXX object CMakeFiles/opmcommon.dir/src/opm/output/data/Solution.cpp.o
[ 55%] Building CXX object CMakeFiles/opmcommon.dir/src/opm/utility/EModel.cpp.o
[ 55%] Linking CXX shared library lib/libopmcommon.so
[ 55%] Built target opmcommon
Scanning dependencies of target test_ExtESmry
Scanning dependencies of target test_ERst
Scanning dependencies of target test_ERft
Scanning dependencies of target test_EclIO
Scanning dependencies of target mocksim
Scanning dependencies of target rewriteEclFile
Scanning dependencies of target test_EclRegressionTest
Scanning dependencies of target test_SummaryNode
Scanning dependencies of target test_Wells
Scanning dependencies of target test_Summary
Scanning dependencies of target test_Inplace
Scanning dependencies of target test_Serializer
[ 55%] Building CXX object CMakeFiles/test_EclIO.dir/tests/test_EclIO.cpp.o
[ 55%] Building CXX object CMakeFiles/test_ExtESmry.dir/tests/test_ExtESmry.cpp.o
[ 55%] Building CXX object CMakeFiles/rewriteEclFile.dir/test_util/rewriteEclFile.cpp.o
[ 55%] Building CXX object CMakeFiles/test_ERst.dir/tests/test_ERst.cpp.o
[ 55%] Building CXX object CMakeFiles/test_SummaryNode.dir/tests/test_SummaryNode.cpp.o
[ 55%] Building CXX object CMakeFiles/test_ERft.dir/tests/test_ERft.cpp.o
[ 55%] Building CXX object CMakeFiles/test_Wells.dir/tests/test_Wells.cpp.o
[ 55%] Building CXX object CMakeFiles/test_Serializer.dir/tests/test_Serializer.cpp.o
[ 55%] Building CXX object CMakeFiles/test_Inplace.dir/tests/test_Inplace.cpp.o
[ 55%] Building CXX object CMakeFiles/mocksim.dir/msim/src/msim.cpp.o
[ 55%] Building CXX object CMakeFiles/test_EclRegressionTest.dir/tests/test_EclRegressionTest.cpp.o
[ 55%] Building CXX object CMakeFiles/test_Summary.dir/tests/test_Summary.cpp.o
[ 55%] Linking CXX executable bin/rewriteEclFile
/bin/ld: lib/libopmcommon.so.2022.04: undefined reference to `PyTuple_SetItem'
/bin/ld: lib/libopmcommon.so.2022.04: undefined reference to `PyObject_Repr'
/bin/ld: lib/libopmcommon.so.2022.04: undefined reference to `PyInstanceMethod_Type'
/bin/ld: lib/libopmcommon.so.2022.04: undefined reference to `PyExc_ValueError'
/bin/ld: lib/libopmcommon.so.2022.04: undefined reference to `PyDict_GetItemString'
/bin/ld: lib/libopmcommon.so.2022.04: undefined reference to `_Py_TrueStruct'
/bin/ld: lib/libopmcommon.so.2022.04: undefined reference to `PyExc_IndexError'
/bin/ld: lib/libopmcommon.so.2022.04: undefined reference to `PyErr_NormalizeException'
/bin/ld: lib/libopmcommon.so.2022.04: undefined reference to `PyInstanceMethod_New'
/bin/ld: lib/libopmcommon.so.2022.04: undefined reference to `PyEval_AcquireThread'
/bin/ld: lib/libopmcommon.so.2022.04: undefined reference to `PyObject_Str'
/bin/ld: lib/libopmcommon.so.2022.04: undefined reference to `PyThreadState_DeleteCurrent'
/bin/ld: lib/libopmcommon.so.2022.04: undefined reference to `PyGILState_GetThisThreadState'
/bin/ld: lib/libopmcommon.so.2022.04: undefined reference to `PyObject_GetAttrString'
/bin/ld: lib/libopmcommon.so.2022.04: undefined reference to `PyCapsule_Type'
[...]
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/test_Summary.dir/build.make:109: bin/test_Summary] Error 1
make[1]: *** [CMakeFiles/Makefile2:823: CMakeFiles/test_Summary.dir/all] Error 2
make: *** [Makefile:160: all] Error 2
```

**Comment by joakim-hove** [comment]
Yes - try to also enable embedded Python in cmake (`OPM_ENABLE_EMBEDDED_PYTHON=ON` in `cmake`) - I agree that should not be necessary - but ....

**Comment by joakim-hove** [comment]
Tip: For the Python/setup.py building it is a great compile time benefit from using `ccache`

**Comment by joakim-hove** [comment]
If you have typed exactly as in the comment the problem is with `DOPM_ENABLE_PYTHON=ON` - which is missing the leading `-`

**Comment by joakim-hove** [comment]
This:
```
bash% cmake ../../opm-common -DOPM_ENABLE_PYTHON=ON -DBUILD_SHARED_LIBS=ON -DOPM_ENABLE_EMBEDDED_PYTHON=ON 
```
works on my machine

**Comment by hakonhagland** [comment]
> which is missing the leading "-"

@joakim-hove  Indeed the leading dash was missing! It compiles now. But when I run the test program `mytest.py` : 

```
$ python3 mytest.py 
Traceback (most recent call last):
  File "/home/hakon/test/opm/opm-common/build4/python/mytest.py", line 1, in <module>
    from opm.io.parser import Builtin
  File "/home/hakon/test/opm/opm-common/build4/python/opm/io/__init__.py", line 1, in <module>
    from .parser import *
  File "/home/hakon/test/opm/opm-common/build4/python/opm/io/parser/__init__.py", line 1, in <module>
    from opm._common import action
  File "/home/hakon/test/opm/opm-common/build4/python/opm/_common.py", line 12, in <module>
    from .libopmcommon_python import action
ImportError: libopmcommon.so.2022.04: cannot open shared object file: No such file or directory
```

**Comment by akva2** [comment]
build static, or set LD_LIBRARY_PATH

**Comment by hakonhagland** [comment]
@akva2 Yes, or maybe we could compile `libopmcommon_python.cpython-39-x86_64-linux-gnu.so` with `-Wl,-rpath` to set an embedded rpath to `libopmcommon.so.2022.04` ?

**Comment by akva2** [comment]
yes, you could do that, you'd have to mess with setup.py(.in) in that case.

**Comment by hakonhagland** [comment]
> Tip: For the Python/setup.py building it is a great compile time benefit from using ccache

@joakim-hove Thanks for the tip! I had actually installed `ccache` and I thought that would be enough. Do I need to specify it when running `cmake` also? Like this maybe: 
```
 cmake -DCMAKE_CXX_COMPILER_LAUNCHER=ccache ..
```

**Comment by joakim-hove** [comment]
There is a small generated script `build/enable-python.sh` which sets the necessary env variables to run the Python version which has been generated in the build directory:
```
bash% source enable-python.sh
```

**Comment by joakim-hove** [comment]
> I had actually installed ccache and I thought that would be enough. Do I need to specify it when running cmake also? Like this maybe:

OK:

1. For using `ccache`withe the normal cmake build you need to set `CMAKE_CXX_COMPILER=ccache ...` as you suggested.
2. For `setup.py` there is basic check to look for `ccache` which should find it automatically.

**Comment by hakonhagland** [comment]
> it is a great compile time benefit from using ccache

@joakim-hove I tested this now:
```
$ mkdir build5
$ cd build5
$ cmake -DOPM_ENABLE_PYTHON=ON -DCMAKE_CXX_COMPILER_LAUNCHER=ccache ..
$ time make -j12
[...]
real	6m22,530s
user	46m50,949s
sys	2m16,846s

$ time make -B -j12
[...]
real	0m12,044s
user	1m27,540s
sys	0m24,533s

$ mkdir build6
$ cmake -DOPM_ENABLE_PYTHON=ON ..
$ time make -j12
real	6m14,784s
user	45m17,688s
sys	1m53,974s

$ time make -B -j12
[...]
real	4m27,585s
user	41m46,392s
sys	1m53,355s
```

So the first compilation takes 8 seconds longer with `ccache` (6m22,530s vs 6m14,784s), whereas a subsequent recompilation of all targets takes 4 minutes and 15 seconds shorter with `ccache` (0m12,044s vs 4m27,585s). So there is a great time saving for this example when using `ccache`. Thanks again!!

**Comment by blattms** [comment]
@hakonhagland Can you retest? Might be fixed by merge of #3238

**Comment by hakonhagland** [comment]
> Can you retest? Might be fixed by merge of https://github.com/OPM/opm-common/pull/3238 or something else

@blattms Yes it works now! Rerunning `make` after changing `DeckKeyword.cpp` recompiles `DeckKeyword.cpp` and regenerates `libopmcommon.a` and `opmcommon_python.cpython-310-x86_64-linux-gnu.so`
